opendc-experiments/opendc-m3sa/src/main/python/models/MultiModel.py

```diff
@@ -0,0 +1,501 @@
+
+        :return: None
+        :side effect: Creates necessary directories and files for output and analysis.
+        """
+        self.output_folder_path = os.getcwd() + "/" + self.path
+        self.raw_output_path = os.getcwd() + "/" + self.path + "/raw-output"
+        self.analysis_file_path = os.getcwd() + "/" + self.path + "/simulation-analysis/"
+        os.makedirs(self.analysis_file_path, exist_ok=True)
+        self.analysis_file_path = os.path.join(self.analysis_file_path, "analysis.txt")
+        if not os.path.exists(self.analysis_file_path):
+            with open(self.analysis_file_path, "w") as f:
+                f.write("Analysis file created.\n")
+
+    def init_models(self):
+        """
+        Initializes models from the simulation output stored in Parquet files. This method reads each Parquet file,
+        processes the relevant data, and initializes Model instances which are stored in the model list.
+
+        :return: None
+        :raise ValueError: If the unit scaling has not been set prior to model initialization.
+        """
+        model_id = 0
+
+        for simulation_folder in os.listdir(self.raw_output_path):
+            if simulation_folder == "metamodel":
+                continue
+            path_of_parquet_file = f"{self.raw_output_path}/{simulation_folder}/seed={self.seed}/{SIMULATION_DATA_FILE}.parquet"
+            parquet_file = pq.read_table(path_of_parquet_file).to_pandas()
+            raw = parquet_file.select_dtypes(include=[np.number]).groupby("timestamp")
+            raw = raw[self.metric].sum().values
+
+            if self.unit_scaling is None:
+                raise ValueError("Unit scaling factor is not set. Please ensure it is set correctly.")
+
+            raw = np.divide(raw, self.unit_scaling)
+
+            if self.user_input["samples_per_minute"] > 0:
+                MINUTES_IN_DAY = 1440
+                self.workload_time = len(raw) * self.user_input["samples_per_minute"] / MINUTES_IN_DAY
+
+            model = Model(raw_sim_data=raw, id=model_id, path=self.output_folder_path)
+            self.models.append(model)
+            model_id += 1
+
+        self.max_model_len = min([len(model.raw_sim_data) for model in self.models])
+
+    def compute_windowed_aggregation(self):
+        """
+        Applies a windowed aggregation function to each model's dataset. This method is typically used for smoothing
+        or reducing data granularity. It involves segmenting the dataset into windows of specified size and applying
+        an aggregation function to each segment.
+
+        :return: None
+        :side effect: Modifies each model's processed_sim_data attribute to contain aggregated data.
+        """
+        if self.plot_type != "cumulative":
+            for model in self.models:
+                numeric_values = model.raw_sim_data
+                model.processed_sim_data = self.mean_of_chunks(numeric_values, self.window_size)
+
+    def generate_plot(self):
```

Comment:
Why would this function not be called with the plot type?
It is not clear to me where this.plot_type is set.

Reply:
Also, I thought the idea was to automatically create multiple plots?

Reply:
well, not really. this depends on the user -- the plotting process takes time as well, right? Why wasting time on plotting if the user may want to only get the data of the metamodel? 

see more in the paper §4, the Meta-Model.

Reply:
the plot_type is set by the user -- either time_series, cumulative, cumulative_time_series.

For more details, have a look at NFRs for §3 and §4 of the paper for better clarification. alternatively, have a look at the documentation.